crates/sui-core/src/transaction_outputs.rs

```diff
@@ -0,0 +1,133 @@
+    ) -> TransactionOutputs {
+        let InnerTemporaryStore {
+            input_objects,
+            mutable_inputs,
+            written,
+            events,
+            max_binary_format_version: _,
+            loaded_runtime_objects: _,
+            no_extraneous_module_bytes: _,
+            runtime_packages_loaded_from_db: _,
+            lamport_version,
+        } = inner_temporary_store;
+
+        let tx_digest = *transaction.digest();
+
+        let deleted: HashMap<_, _> = effects.all_tombstones().into_iter().collect();
+
+        // Get the actual set of objects that have been received -- any received
+        // object will show up in the modified-at set.
+        let modified_at: HashSet<_> = effects.modified_at_versions().into_iter().collect();
+        let possible_to_receive = transaction.transaction_data().receiving_objects();
+        let received_objects = possible_to_receive
+            .iter()
+            .cloned()
+            .filter(|obj_ref| modified_at.contains(&(obj_ref.0, obj_ref.1)));
+
+        // We record any received or deleted objects since they could be pruned, and smear shared
+        // object deletions in the marker table. For deleted entries in the marker table we need to
+        // make sure we don't accidentally overwrite entries.
+        let markers: Vec<_> = {
+            let received = received_objects
+                .clone()
+                .map(|objref| (ObjectKey::from(objref), MarkerValue::Received));
+
+            let deleted = deleted.into_iter().map(|(object_id, version)| {
+                let object_key = ObjectKey(object_id, version);
+                if input_objects
+                    .get(&object_id)
+                    .is_some_and(|object| object.is_shared())
+                {
+                    (object_key, MarkerValue::SharedDeleted(tx_digest))
+                } else {
+                    (object_key, MarkerValue::OwnedDeleted)
+                }
+            });
+
+            // We "smear" shared deleted objects in the marker table to allow for proper sequencing
+            // of transactions that are submitted after the deletion of the shared object.
+            // NB: that we do _not_ smear shared objects that were taken immutably in the
+            // transaction.
+            let smeared_objects = effects.deleted_mutably_accessed_shared_objects();
+            let shared_smears = smeared_objects.into_iter().map(move |object_id| {
+                (
+                    ObjectKey(object_id, lamport_version),
+                    MarkerValue::SharedDeleted(tx_digest),
+                )
+            });
+
+            received.chain(deleted).chain(shared_smears).collect()
+        };
+
+        let locks_to_delete: Vec<_> = mutable_inputs
+            .into_iter()
+            .filter_map(|(id, ((version, digest), owner))| {
+                owner.is_address_owned().then_some((id, version, digest))
+            })
+            .chain(received_objects)
```

Comment:
I  see that you are doing a little code optimizations here by merging the two lists so that you could delete the locks altogether. But it's optimizations like this that make review tricky.. I had a hard time reason about this change. I think code cleanups (i.e. consolidating statements, optimizations and etc.) should always be in separate PRs from mechanical API refactoring, especially when you are moving chunks of code from one place to another, otherwise it's so easy to miss issues.
I think a good rule of thumb is that when doing large mechanical API refactoring all logic should stay the same, even if you don't like them, and clean them up in latter PRs..

Reply:
fair enough! I don't even recall doing this, must not have been thinking